include/ginkgo/core/base/mpi.hpp

```diff
@@ -1308,12 +1308,38 @@ class communicator {
                       const SendType* send_buffer, const int* send_counts,
                       const int* send_offsets, RecvType* recv_buffer,
                       const int* recv_counts, const int* recv_offsets) const
+    {
+        this->all_to_all_v(std::move(exec), send_buffer, send_counts,
+                           send_offsets, type_impl<SendType>::get_type(),
+                           recv_buffer, recv_counts, recv_offsets,
+                           type_impl<RecvType>::get_type());
+    }
+
+    /**
+     * Communicate data from all ranks to all other ranks with
+     * offsets (MPI_Alltoallv). See MPI documentation for more details.
+     *
+     * @param exec  The executor, on which the message buffers are located.
+     * @param send_buffer  the buffer to send
+     * @param send_count  the number of elements to send
+     * @param send_offsets  the offsets for the send buffer
+     * @param send_type  the MPI_Datatype for the send buffer
+     * @param recv_buffer  the buffer to gather into
+     * @param recv_count  the number of elements to receive
+     * @param recv_offsets  the offsets for the recv buffer
+     * @param recv_type  the MPI_Datatype for the recv buffer
+     * @param comm  the communicator
+     */
+    void all_to_all_v(std::shared_ptr<const Executor> exec,
```

Comment:
I think it should be like the above version using SendType and RecvType?
also scaled the offset by the type

Reply:
Could you expand on in which way this should be like the templated one? This one is needed for the custom MPI_Datatype to handle multiple right-hand-sides.

Reply:
I think that is an MPI detail that we don't need to care about.

Reply:
But we don't do anything here, just pass everything on to MPI. So the user just has to provide it so that it matches the MPI use.

Reply:
we have two version for that
1. with the templated type -> it will add the MPItype and then pass to the following option
2. with the void type 

From my view on this interface, I will pass the offset like (1, 2, 3) no matter what type is because it takes care of the type. However, I will consider to pass the offset like (2, 4, 6) when the type size is 8 bytes for the void version because I do not know whether the function will take care of the type size for the offset.

After reading some documentation and the implementation of 1, it seems to take care of the type size. The alltoallw is another version which only consider the offset in bytes. 

It does not need to add the template on the void version but could you add some short documentation in void version?

Reply:
How the offset is interpreted depends on the MPI_Datatype, I think you can read up on that somewhere in the MPI standard. So the offset is not in byte, but something depending on the datatype. 
All of these functions are only wrapper for the corresponding MPI functions, so I don't think we should add documentation here. The only useful documentation I could think of is referring to the MPI standard, but that seems superfluous.

Reply:
TBH, the documentation in MPI is not clear to me. At least from MPICH, they only specify MPI_Alltoallw is in byte, but not mention specific in MPI_Alltoallv. Without looking into MPI_alltoallw, I will not think MPI_alltoallv might using element-wise.
Okay, I found more clear document in [openmpi](https://www.open-mpi.org/doc/v4.0/man3/MPI_Alltoallv.3.php) and [rookiehpc](https://rookiehpc.github.io/mpi/docs/mpi_alltoallv/index.html) not like [MPICH](https://www.mpich.org/static/docs/v3.1/www3/MPI_Alltoallv.html). they specify it is in units of sendtype
I mean in the param doc not the function description.
something like `the element-wise offsets for the send buffer` or `the offsets (in units of send_type) for the send buffer` for `send_offsets` in void version
If you think it is still confusing or might not match the mpi document, you can keep the original version.

Reply:
We don't explain the MPI functions here, and I don't think it would be our job to do so. One option would be to remove the parameter description on all of our MPI wrapper and just keep the reference to the MPI standard. Would you be on board with that?

Reply:
I already elaborated a bit on that on slack, but I think this function should at the very least have a different name, preferably be more specialized to allow making it type-safe instead of `void*`-typed and resolving the whole offset question.